File: sdk/php/src/Http/AuthResult.php
<?php

declare(strict_types=1);

namespace Xid\Http;

use Xid\Jwt\Claims;

/**
 * 请求认证结果值对象。
 *
 * 使用判别联合风格:通过 isAuthenticated() 判断状态,
 * 已认证时调用 claims(),未认证时调用 reason()。
 */
final class AuthResult
{
    private function __construct(
        private readonly bool $authenticated,
        private readonly Claims|null $claims,
        private readonly string|null $reason,
    ) {}

    /**
     * 构造已认证结果。
     */
    public static function authenticated(Claims $claims): self
    {
        return new self(true, $claims, null);
    }

    /**
     * 构造未认证结果。
     *
     * @param string $reason  人类可读的失败原因(不对外暴露细节,仅供服务端日志)
     */
    public static function unauthenticated(string $reason): self
    {
        return new self(false, null, $reason);
    }

    public function isAuthenticated(): bool
    {
        return $this->authenticated;
    }

    /**
     * 返回已验证的 claims。
     * 调用前必须确认 isAuthenticated() === true。
     *
     * @throws \LogicException  若未认证时调用
     */
    public function claims(): Claims
    {
        if ($this->claims === null) {
            throw new \LogicException('Cannot access claims on an unauthenticated result');
        }
        return $this->claims;
    }

    /**
     * 返回失败原因字符串(未认证时才有值)。
     */
    public function reason(): string|null
    {
        return $this->reason;
    }
}

File: sdk/java/src/main/java/dev/xid/sdk/AuthResult.java
package dev.xid.sdk;

import java.util.Optional;

/**
 * 请求认证结果,由 {@link XidClient#authenticateRequest} 返回。
 *
 * 模式:
 *   AUTHENTICATED   -- token 有效,claims 已填充
 *   UNAUTHENTICATED -- 无 token(Authorization header 缺失 / cookie 不存在)
 *   INVALID         -- token 存在但验证失败,reason 描述原因
 *
 * 典型用法:
 * <pre>{@code
 * AuthResult result = client.authenticateRequest(request);
 * if (result.isAuthenticated()) {
 *     String userId = result.getClaims().getSub();
 *     // ...
 * } else {
 *     // 返回 401
 * }
 * }</pre>
 */
public final class AuthResult {

    public enum Status {
        AUTHENTICATED,
        UNAUTHENTICATED,
        INVALID
    }

    private final Status     status;
    private final XidClaims  claims;
    private final String     reason;

    private AuthResult(Status status, XidClaims claims, String reason) {
        this.status = status;
        this.claims = claims;
        this.reason = reason;
    }

    /** 认证成功 */
    static AuthResult authenticated(XidClaims claims) {
        return new AuthResult(Status.AUTHENTICATED, claims, null);
    }

    /** 无 token */
    static AuthResult unauthenticated() {
        return new AuthResult(Status.UNAUTHENTICATED, null, null);
    }

    /** token 存在但验证失败 */
    static AuthResult invalid(String reason) {
        return new AuthResult(Status.INVALID, null, reason);
    }

    public Status getStatus() {
        return status;
    }

    public boolean isAuthenticated() {
        return status == Status.AUTHENTICATED;
    }

    public boolean isUnauthenticated() {
        return status == Status.UNAUTHENTICATED;
    }

    /**
     * 验证成功时的 claims。
     * status != AUTHENTICATED 时返回 {@link Optional#empty()}。
     */
    public Optional<XidClaims> getClaims() {
        return Optional.ofNullable(claims);
    }

    /**
     * INVALID 时的失败原因描述,其余情况返回 null。
     */
    public String getReason() {
        return reason;
    }

    @Override
    public String toString() {
        return "AuthResult{status=" + status + ", reason=" + reason + "}";
    }
}
